numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1413,19 +1507,19 @@ PyUFunc_DefaultLegacyInnerLoopSelector(PyUFuncObject *ufunc,
         types += nargs;
     }
 
-    errmsg = PyUString_FromFormat("ufunc '%s' did not contain a loop "
-                    "with signature matching types ", ufunc_name);
-    for (i = 0; i < nargs; ++i) {
-        PyUString_ConcatAndDel(&errmsg,
-                PyObject_Repr((PyObject *)dtypes[i]));
-        if (i < nargs - 1) {
-            PyUString_ConcatAndDel(&errmsg, PyUString_FromString(" "));
+    /* Produce an error object */
+    {
+        /* convert dtypes to a tuple */
```

Comment:
I'd move all of this to `raise_no_loop_found_error` as well (even gets rid of the no-clause curly brackets!)